lib/kubernetes-deploy/resource_watcher.rb

```diff
@@ -11,13 +11,18 @@ def initialize(resources, logger:, deploy_started_at: Time.now.utc, operation_na
       @logger = logger
       @deploy_started_at = deploy_started_at
       @operation_name = operation_name
+      @timeout = timeout
     end
 
     def run(delay_sync: 3.seconds, reminder_interval: 30.seconds, record_summary: true)
-      delay_sync_until = last_message_logged_at = Time.now.utc
+      delay_sync_until = last_message_logged_at = monitoring_started = Time.now.utc
       remainder = @resources.dup
 
       while remainder.present?
+        if @timeout && (Time.now.utc - monitoring_started > @timeout)
+          report_giving_up(remainder)
+          break
```

Comment:
I'd like to`raise DeploymentTimeoutError, failed_resources` instead of `break` and do any desired reporting in `report_giving_up`. I think that's cleaner than modifying the resources. 

👍 to (3)